### PR TITLE
fix: (platform) upload collection

### PR DESCRIPTION
**Pull request by Betrozov**
#### Please provide a link to the associated issue.
https://github.com/SAP/fundamental-ngx/issues/3171
#### Please provide a brief summary of this pull request.
Introduce new Platform Upload Collection component
#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [X] Documentation Examples
- [X] Stackblitz works for all examples

### CODE REVIEW

File: libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.ts [hunk 1]
```diff
@@ -0,0 +1,1005 @@
...
+    MessageStripType
+} from '../models/upload-collection.models';
+import {
+    TypeMissmatchEvent,
+    FilenameLengthExceedEvent,
+    FileSizeExceedEvent,
+    UploadEventResponse,
+} from '../models/upload-collection-events.models';
+import { generateMessageStripeData, generateUuid } from '../helpers';
+import { take } from 'rxjs/operators';
```
Comment (IlyaSurmay) [inline libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.ts:28]: please move this up to lib imports

File: libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.ts [hunk 2]
```diff
@@ -0,0 +1,1005 @@
...
+                    .subscribe(newDataSource => {
+                        this._dataSource = newDataSource;
+                        this.showMessage(CreateMessageType.MOVE, {
+                            payload: payload,
+                            type: MessageStripType.SUCCESS
+                        });
+                    });
+            },
+            (error) => {
+                console.log(error);
```
Comment (IlyaSurmay) [inline libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.ts:431]: maybe `console.error(error)` is more suitable here?

**Review by IlyaSurmay — CHANGES_REQUESTED** [review]

File: apps/docs/src/app/platform/component-docs/upload-collection/upload-collection-header/upload-collection-header.component.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'fd-upload-collection-header',
+  templateUrl: './upload-collection-header.component.html',
+  styleUrls: ['./upload-collection-header.component.scss']
+})
+export class UploadCollectionHeaderComponent implements OnInit {
+
+  constructor() { }
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply apps/docs/src/app/platform/component-docs/upload-collection/upload-collection-header/upload-collection-header.component.ts]: fixed

File: libs/platform/src/lib/components/upload-collection/dialogs/move-to/move-to.component.html [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                        compact="true"
+                        label="New Folder"
+                        (click)="newFolder()"
+                    ></button>
+                </fd-bar-element>
+            </div>
+        </ng-template>
+    </fd-dialog-header>
+
+    <fd-dialog-body style="padding-left: 0; padding-right: 0">
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/dialogs/move-to/move-to.component.html]: fixed

File: libs/platform/src/lib/components/upload-collection/directives/upload-collection-dragndrop.directive.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    @Output()
+    readonly onDropzone = new EventEmitter<boolean>();
+
+    /** @hidden */
+    private dropzonePreviousState: boolean;
+
+    /** @hidden */
+    private elementSelector: string;
+
+    // .fdp-upload-collection__table
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/directives/upload-collection-dragndrop.directive.ts]: removed

File: package.json [hunk 6]
```diff
@@ -37,7 +37,7 @@
     "ng": "ng",
     "release": "./scripts/publish-release.sh",
     "release:create": "create-release",
-    "start": "ng serve",
+    "start": "node --max_old_space_size=12288 ./node_modules/@angular/cli/bin/ng serve",
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply package.json]: reverted

File: libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.ts (hunk 1 above)
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.ts]: sorted

File: libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.ts (hunk 2 above)
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.ts]: changed

**Review by IlyaSurmay — APPROVED** [review]

File: apps/docs/src/app/platform/component-docs/upload-collection/examples/upload-collection-base.service.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    UploadCollectionService,
+    UploadEvent,
+    UploadEventResponse
+} from '@fundamental-ngx/platform';
+
+@Injectable()
+export class UploadCollectionBaseService extends UploadCollectionService {
+    /** The method is triggered when valid files are selected in the file uploader dialog. */
+    upload(data: UploadEvent): Observable<UploadEventResponse[]> {
+        console.log('upload', data);
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply apps/docs/src/app/platform/component-docs/upload-collection/examples/upload-collection-base.service.ts:24]: This is just an example of how to work with `UploadCollectionService`.
All the consoles need here to see the result in the browser console

File: libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  beforeEach(() => {
+    fixture = TestBed.createComponent(UploadCollectionComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+});
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/upload-collection/upload-collection.component.spec.ts:25]: unit tests will be tomorrow

File: libs/platform/src/lib/components/upload-collection/dialogs/new-folder/new-folder.component.html [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    <fd-dialog-footer>
+        <fd-dialog-footer-button>
+            <button
+                fd-button
+                fdType="emphasized"
+                fd-dialog-decisive-button
+                label="Create"
+                [compact]="true"
+                (click)="this.dialogRef.close(newFolderName)"
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/dialogs/new-folder/new-folder.component.html:30]: fixed

File: libs/platform/src/lib/components/upload-collection/dialogs/new-folder/new-folder.component.html [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+        <fd-dialog-footer-button>
+            <button
+                fd-button
+                fdType="transparent"
+                fd-dialog-decisive-button
+                fd-initial-focus
+                label="Cancel"
+                [compact]="true"
+                (click)="this.dialogRef.dismiss()"
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/dialogs/new-folder/new-folder.component.html:42]: fixed

File: libs/platform/src/lib/components/upload-collection/dialogs/move-to/move-to.component.html [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                Move
+            </button>
+        </fd-dialog-footer-button>
+        <fd-dialog-footer-button>
+            <button
+                fd-button
+                fdType="transparent"
+                fd-dialog-decisive-button
+                compact="true"
+                (click)="dialogRef.dismiss()"
```
[the remark this answers is not in the document]
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/dialogs/move-to/move-to.component.html:75]: fixed

[the remark this answers is not in the document]
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/dialogs/move-to/move-to.component.html:75]: fixed

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (olshansky) [reply package.json:40]: @Betrozov it's still here 🤔

**Comment by fkolar** [comment]
My Major comment would be that `Upload Collection `should be part of the Inputs. Upload components are usually used as part of the forms. you upload attachments to existing contracts and this is saved into the model (Contract documents etc.. )

We shoudl support Form API, To be able to validate , show error messages as part of the forms, etc.

**Comment by Betrozov** [comment]
> My Major comment would be that `Upload Collection `should be part of the Inputs. Upload components are usually used as part of the forms. e.g.: you upload attachments to existing contracts and this is saved into the model (Contract documents etc.. ), and soo on.
> 
> We shoudl support Form API, To be able to validate , show error messages as part of the forms, etc.

Hi @fkolar  :slightly_smiling_face: 

Regarding your comment, I have not seen in the docs any info about that...

While working on this component I have based on the following:
- https://sapui5.hana.ondemand.com/#/entity/sap.m.UploadCollection
- https://experience.sap.com/fiori-design-web/upload-collection/
- UC grid table - Ariba design

The listed sources did mention nothing about usage with Form.
`sapUI5` examples based on the instant upload strategy as well

**Comment by Betrozov** [comment]
> I probably missed the review, otherwise I woudl make the change right there,
> we dont need Inputs like
> 
> uploadedBy, as this does not make sense on this Platform Level thsi is more Application specific behavior. Also I see in docs that there is binding called

`uploadedBy` - removed

> noDataText, -> Let;s sync it up with other components that we already have.
> In Table we use nodataMessage, some other place we have noDataLabel, etc..

The `UploadCollectionComponent` uses the same Input name `noDataText` as `BaseListItem` directive

**Comment by fkolar** [comment]
I think its implicit. They dont need to mention this right? Let's try to think in what context is this Upload or UploadCollection is going to be used? Can you list usecases? 


I can see on the link you sent me that There is system to track tasks, Or Purchasing system, Contract Management and you are always uploading files within the context of some model. 

So you in real usecase you will have a page with different form fields and there will be also place to upload some attachemnts, and all this you should be able to link to one ENTITY (Domain Model instance) and then send all this togehter to the backend. 

So the rest of the field values from e.g. Contract will be automatically synced with Object, but in case of uploaded files application developer will have to do some addiional tasks.

**Comment by fkolar** [comment]
We shoudl try to more understand how this component is going to be used in what realistic example.  Adding kevin to this discussion. To bad i mised the review on gitbook

**Comment by Betrozov** [comment]
I just want to be on the same page with you.
So, we are talking about Ariba Design, Upload Collection use the Grid table as a container.
upload-collection
FYI: @fkolar @KevinOkamoto

**Comment by KevinOkamoto** [comment]
I'm unsure about @fkolar comment that the Upload Collection should be an input component within a form. My initial thoughts on the Upload Collection based on the UX mock-ups, is that this is a stand-alone complex component, similar to Approval Flow. Because the Upload Collection encapsulates all the CRUD interfaces in one component, I felt that it would need to interface with a specific file management service, not as part of a general domain model service. However I could be wrong.

Perhaps we should do more research, and ask PMs and UX designers under what context this component will be used. For now can we go with the current design, and if we need to change this component to be an Input component we can open a new PR for that?

**Comment by Betrozov** [comment]
> If it handles its own CRUD operation then I am ok, but I did not see any CRUD operation in implemenation, it woudl be up to the developer to handle this.

Hi @fkolar 
plz look at the [UploadCollectionService](https://github.com/valor-software/fundamental-ngx/pull/95/files#diff-353c55c41e928d6403f0186911ca300e8748f253aa0667e0ebeb4ad301d19888R20)

**Comment by fkolar** [comment]
> > If it handles its own CRUD operation then I am ok, but I did not see any CRUD operation in implemenation, it woudl be up to the developer to handle this.
> 
> Hi @fkolar
> plz look at the [UploadCollectionService](https://github.com/valor-software/fundamental-ngx/pull/95/files#diff-353c55c41e928d6403f0186911ca300e8748f253aa0667e0ebeb4ad301d19888R20)

Good. Even the idea of having dataSOURCE is that dataSOURCE uses DataProvider and data Provider knows how to handle the specific types of data. 

DataProviders is responsible for handling CRUD operation

**Comment by Betrozov** [comment]
@fkolar please review one more time

File: libs/platform/src/lib/components/upload-collection/directives/upload-collection-dragndrop.directive.ts [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+import { FilesValidatorService, FilesValidatorOutput } from '../services';
+
+/**
+ * Directive that handles the drag and drop feature of the file input.
+ */
+@Directive({
+    selector: '[fdpUploadCollectionDragnDrop]'
+})
+export class UploadCollectionDragndropDirective {
```
Comment (fkolar) [inline libs/platform/src/lib/components/upload-collection/directives/upload-collection-dragndrop.directive.ts:13]: Let's syncup the name with the directive selector. 

DragnDrop vs Dragndrop

File: apps/docs/src/app/platform/component-docs/upload-collection/examples/upload-collection-base-data-provider.ts [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+
+    /** The method is triggered when Update Version button is pressed and valid file are selected in the file uploader dialog. */
+    updateVersion({ parentFolderId, item, newItem }: UpdateVersionEvent): Observable<UploadCollectionItem[]> {
+        console.log('updateVersion', parentFolderId, item);
+
+        let updatedItem: UploadCollectionFile = {
+            ...item,
+            status: UploadCollectionItemStatus.SUCCESSFUL,
+            uploadedOn: new Date(),
+            uploadedBy: {
```
Comment (fkolar) [inline apps/docs/src/app/platform/component-docs/upload-collection/examples/upload-collection-base-data-provider.ts:159]: uploadedBy are these some default value.

**Review by fkolar — CHANGES_REQUESTED** [review]

File: libs/platform/src/lib/components/upload-collection/directives/upload-collection-dragndrop.directive.ts (hunk 12 above)
Comment (Betrozov) [reply libs/platform/src/lib/components/upload-collection/directives/upload-collection-dragndrop.directive.ts]: fixed

File: apps/docs/src/app/platform/component-docs/upload-collection/examples/upload-collection-base-data-provider.ts (hunk 13 above)
Comment (Betrozov) [reply apps/docs/src/app/platform/component-docs/upload-collection/examples/upload-collection-base-data-provider.ts:159]: ignored

**Review by fkolar — APPROVED** [review]

**Pull request closed without merging**